### PR TITLE
 Catchup, replicate and Merge segments

### PR DESCRIPTION
**Pull request by zalokhan**
### Workflow: Merge segment

This workflow is executed only if the new node to be added is going to be a log unit server.
Steps of these workflow are dependent on the type of replication protocol being used. The current implementation strongly assumes Chain replication only. Abstraction to support other replication modes might be missing.
Comprises of 3 steps:
- Step 1: Catch-up segment.
  - This step fetches a list of all known addresses from the existing head and tail of the chain.
  - The diff of all the addresses between these 2 sets is calculated.
  - A read is performed (on the tail) for all the missing addresses (diff set) which forces hole fills.

- Step 2: Segment replication
  - At the source (any node in the existing chain as all nodes have the same state after catch-up) the 
    files are copied into a buffer and sent to the destination node. (No lock or snapshot is taken on 
    the file being copied)
  - At the destination, the file is opened by a temporary segment handler and copied to the existing 
    file to preserve existing data.

- Step 3: Merge Segment
  - The layout is now reconfigured to merge the last 2 segments.

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **9e40dc7**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #971 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/11Laas5LJh5McOAvHHeOjQ6htBx4DACQ)

**Comment by Maithem** [comment]
This is no longer required, correct?

**Comment by zalokhan** [comment]
This is required.
This follows #970 Split segment and add node.
This PR contains the data replication and merge segment.
I closed the irrelevant PRs

File: infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java [hunk 1]
```diff
@@ -286,6 +287,20 @@ private void flushCache(CorfuMsg msg, ChannelHandlerContext ctx, IServerRouter r
         r.sendResponse(ctx, msg, CorfuMsgType.ACK.msg());
     }
 
+    /**
+     * Handles requests for replicating
```
Comment (Maithem) [inline infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java]: We can't use this, the time it takes to transfer a segment is equal to the time it takes to write it to the address space. In other words, its slow. Also, there are many corner cases that are not considered by the replicateRawData implementation, for example the interleaving of trimmed addresses and non-trimmed addresses. 

I re-implemented it here: https://github.com/CorfuDB/CorfuDB/pull/1004/files

**Comment by zalokhan** [comment]
Removing state transfer code as it will use batch writes.
Merging the rest of the "merge segments" code to #970

**Pull request closed without merging**